File: golang/_test/check_string_test.go
package _test

import (
	"github.com/stretchr/testify/require"
	"str/utils"
	"testing"
)

func TestCheckIsEmail(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		shouldBe bool
	}{
		{
			name:     "_test 1",
			input:    "admin@gmail.com",
			shouldBe: true,
		},
		{
			name:     "_test 2",
			input:    "student@student.univ.ac.id",
			shouldBe: true,
		},
		{
			name:     "_test 3",
			input:    "not a email",
			shouldBe: false,
		},
		{
			name:     "_test 4",
			input:    "asdfg@asdf.asd",
			shouldBe: false,
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			result := utils.IsEmail(test.input)
			require.Equal(t, test.shouldBe, result)
		})
	}
}

func BenchmarkCheckIsEmail(b *testing.B) {
	tests := []struct {
		name  string
		input string
	}{
		{
			name:  "_test 1",
			input: "admin@gmail.com",
		},
		{
			name:  "_test 2",
			input: "student@student.univ.ac.id",
		},
		{
			name:  "_test 3",
			input: "not a email",
		},
		{
			name:  "_test 4",
			input: "asdfg@asdf.asd",
		},
	}

	for _, test := range tests {
		b.Run(test.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				utils.IsEmail(test.input)
			}
		})
	}
}

func TestCheckSimpleRole(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		shouldBe string
	}{
		{
			name:     "_test 1",
			input:    "U",
			shouldBe: "users",
		},
		{
			name:     "_test 2",
			input:    "A",
			shouldBe: "admin",
		},
		{
			name:     "_test 3",
			input:    "SA",
			shouldBe: "super admin",
		},
		{
			name:     "_test 4",
			input:    "u",
			shouldBe: "unknown",
		},
		{
			name:     "_test5",
			input:    "MA",
			shouldBe: "unknown",
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			result := utils.SimpleCheckRole(test.input)
			require.Equal(t, test.shouldBe, result)
		})
	}
}

func BenchmarkCheckSimpleRole(b *testing.B) {
	tests := []struct {
		name     string
		input    string
		shouldBe string
	}{
		{
			name:  "_test 1",
			input: "U",
		},
		{
			name:  "_test 2",
			input: "A",
		},
		{
			name:  "_test 3",
			input: "SA",
		},
		{
			name:  "_test 4",
			input: "u",
		},
		{
			name:  "_test5",
			input: "MA",
		},
	}

	for _, test := range tests {
		b.Run(test.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				utils.SimpleCheckRole(test.input)
			}
		})
	}
}

File: rust/src/test/check_string_test.rs
struct TestIsEmailStruct {
    str: String,
    should_be: bool,
}

use crate::utils::check_string::{is_email, simple_check_role};

#[test]
fn test_check_is_email() {
    let tests: [TestIsEmailStruct; 4] = [
        TestIsEmailStruct{ str: String::from("admin@gmail.com"), should_be: true },
        TestIsEmailStruct{ str: String::from("student@student.univ.ac.id"), should_be: true },
        TestIsEmailStruct{ str: String::from("not a email"), should_be: false },
        TestIsEmailStruct{ str: String::from("asdfg@asdf.asd"), should_be: false },
    ];

    for test in tests {
        let result = is_email(&test.str);
        assert_eq!(result, test.should_be)
    }
}

struct TestCheckRoleStruct {
    str: String,
    should_be: String,
}

#[test]
fn test_simple_check_role() {
    let tests: [TestCheckRoleStruct; 5] = [
        TestCheckRoleStruct{ str: String::from("U"), should_be: String::from("users") },
        TestCheckRoleStruct{ str: String::from("A"), should_be: String::from("admin") },
        TestCheckRoleStruct{ str: String::from("SA"), should_be: String::from("super admin") },
        TestCheckRoleStruct{ str: String::from("u"), should_be: String::from("unknown") },
        TestCheckRoleStruct{ str: String::from("MR"), should_be: String::from("unknown") },
    ];
    for test in tests {
        let result = simple_check_role(&test.str);
        assert_eq!(result, test.should_be)
    }
}
